src/puppetlabs/puppetdb/threadpool.clj

```diff
@@ -0,0 +1,130 @@
+  down the threadpool that work to complete"
+  [{:keys [^Semaphore semaphore ^ExecutorService threadpool shutdown-timeout]}]
+
+  ;; Prevent new work from being accepted
+  (.drainPermits semaphore)
+
+  ;; Shutdown the threadpool, allowing in-flight work to finish
+  (.shutdown threadpool)
+
+  ;; This will block, waiting for the threadpool to shutdown
+  (when-not (.awaitTermination threadpool shutdown-timeout java.util.concurrent.TimeUnit/MILLISECONDS)
+
+    (log/warn
+     (i18n/trs "Threadpool not stopped after {0} milliseconds, forcibly shutting it down" shutdown-timeout))
+
+    ;; This will force the shutdown of the threadpool and will
+    ;; not allow current threads to finish
+    (.shutdownNow threadpool)
+    (log/warn
+     (i18n/trs "Threadpool forcibly shutdown"))))
+
+(defrecord GatedThreadpool [^Semaphore semaphore ^ExecutorService threadpool shutdown-timeout]
+  java.io.Closeable
+  (close [this]
+    (shutdown this)))
+
+(defn create-threadpool
+  "Creates an unbounded threadpool with the intent that access to the
+  threadpool is bounded by the semaphore. Implicitly the threadpool is
+  bounded by `size`, but since the semaphore is handling that aspect,
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size name-pattern shutdown-timeout-in-ms]
+  (let [threadpool (ThreadPoolExecutor. 1
+                                        Integer/MAX_VALUE
+                                        1
+                                        TimeUnit/MINUTES
+                                        (SynchronousQueue.))]
+    (->> threadpool
+         .getThreadFactory
+         (thread-factory name-pattern)
+         (.setThreadFactory threadpool))
+
+    (map->GatedThreadpool
+     {:semaphore (Semaphore. size)
+      :threadpool threadpool
+      :shutdown-timeout shutdown-timeout-in-ms})))
+
+(defn call-on-threadpool
+  "Executes `f` on a gated threadpool. Calls `on-complete` with the
+  results of inoking `f`"
+  [{:keys [^Semaphore semaphore ^ExecutorService threadpool] :as gated-threadpool}
+   f]
+  (.acquire semaphore)
+  (try
+    (.execute threadpool (fn []
+                           (try
+                             (f)
+                             (finally
+                               (.release semaphore)))))
+    ;; RejectedExecutionExceptions only occur when attempting ot
+    ;; excecute new work on a threadpool that is in the process of
+    ;; shutting down or already has been shut down
+    (catch RejectedExecutionException e
+      (.release semaphore)
+      (throw e))))
+
+(defn exec-from-channel
+  "Executes `on-input` for each input found on `in-chan`, with a
+  threadpool of size `parallelism`. The results of the `on-input`
+  invocations will be put to the `out-chan` in the order it has
+  completed. Throws when the threadpool is shutting down and there are
+  still messages in the channel"
+  [gated-threadpool in-chan on-input]
```

Comment:
This is a small, picky note but this function reminds me a lot of `map` and it might be nice to have the signature be `[gated-threadpool f chan]`. Also the docstring is wrong now I think because `out-chan` isn't a thing.